crates/sparrow-journal/src/lib.rs

```diff
@@ -0,0 +1,403 @@
+        // process of checkpointing.
+        //
+        // This may cause new inputs to pause while checkpoints are taken.
+        // If this becomes a problem, we could look at a more sophisticated
+        // strategy -- for instace, we could first take the batches and concatenate
+        // them into a "pending-checkpoint", at which point we could release the lock.
+        // Then, once the checkpoint completes, we could acquire the lock and drop the
+        // pending checkpoint batch (since it is now available in the checkpoint).
+        let mut lock = self.batches.lock();
+        let batches = std::mem::take(&mut lock.0);
+        self.checkpoints.lock().checkpoint(batches, lock.1)?;
+        Ok(())
+    }
+
+    fn in_memory_stream(
+        &self,
+    ) -> error_stack::Result<BoxStream<'static, error_stack::Result<Batch, Error>>, Error> {
+        // This clones the vec to create an owned iterator without needing to hold the lock.
+        // Cloning the vec (and batches) shouldn't be too problematic given their expected
+        // size. If they are, we could concatenate them into a single batch, updating the
+        // in-memory set and then only cloning that one batch.
+        Ok(futures::stream::iter(self.batches.lock().0.clone())
+            .map(Ok)
+            .boxed())
+    }
+
+    /// Return a stream over the current contents.
+    fn stream_current(
+        &self,
+    ) -> error_stack::Result<BoxStream<'static, error_stack::Result<Batch, Error>>, Error> {
+        // Since we don't currently accept late data, we don't need to merge with the checkpoints.
+        // Instead, we can just chain at the end.
+        let checkpoint_stream = self.checkpoints.lock().read_checkpoints()?;
+        let in_memory_stream = self.in_memory_stream()?;
+        Ok(checkpoint_stream.chain(in_memory_stream).boxed())
+    }
+
+    /// Return a stream over the current contents and subscribed to future additions.
+    fn stream_subscribe(
+        &self,
+    ) -> error_stack::Result<BoxStream<'static, error_stack::Result<Batch, Error>>, Error> {
+        // TODO: Merge with checkpoints, and subscribe to future changes.
+        self.stream_current()
+    }
+}
+
+#[derive(Debug)]
+struct BatchCheckpointer {
+    batches: Arc<Batches>,
+    decoder: BatchDecoder,
+}
+
+impl okaywal::LogManager for BatchCheckpointer {
+    fn recover(&mut self, entry: &mut okaywal::Entry<'_>) -> std::io::Result<()> {
+        // DO NOT SUBMIT: Handle errors
+        if let Some(batch) = self.decoder.decode(entry).unwrap() {
+            self.batches.add_batch(batch).unwrap();
+        }
+        Ok(())
+    }
+
+    fn checkpoint_to(
+        &mut self,
+        _last_checkpointed_id: okaywal::EntryId,
+        _checkpointed_entries: &mut okaywal::SegmentReader,
+        _wal: &okaywal::WriteAheadLog,
+    ) -> std::io::Result<()> {
+        // TODO: This should use `_last_checkpointed_id` to ensure that only batches
```

Comment:
I think this is probably critical, and it may reveal an issue with how things are structured (and how we anticipated implementing them).

Specifically:

1. The checkpointing happens asynchronously. So we *should not* just use the current in-memory batches for checkpointing.
2. This means that some batches may have been added to the new segment (and thus the "in-memory" batch vector) since we decided to checkpoint. These batches *should not* be part of the checkpoint (they'll be in the next one) or we'll have duplicate rows.

Possible options:

1. De-duplicate if rows show up in the checkpoint and segment.
2. When checkpointing, use the entries themselves, rather than the "in-memory batches" -- this wouldn't work since the in-memory batches need to be flushed at some point.
3. Use some kind of "timestamped in-memory batches" (associated with the entry IDs). It shouldn't be difficult to associate each element in the vector with the corresponding `EntryId`, and only take a prefix of the vector corresponding to those entries.

I suspect that case 3 is the best option. While we're not trying to handle late data, it does mean that the late data strategy gets a little trickier. We have a few options:

A) When late data arrives, force a checkpoint. Simple, but in the presence of lots of late data this would lead to slow-down and excessively small checkpoints.
B) Leave the batches in memory unordered. When we need to execute on them, we can check to see if there is any overlap, and decide to do (i) concatenation if there is none or (ii) concatenation and sorting if there is or (iii) sorting and merging.
C) Something more intelligent, that manages both the entry IDs associated with batches and the late data.

I suspect B may be the best option, unless we have a way of detecting when a checkpoint is taken.

Reply:
Agreed, critical to do. And agreed with option 3 -- disregarding late data issues for now, the checkpointer could keep track of the timestamps for each checkpoint, and have separate vectors for each set of batches. Or, we could include a `checkpoint_version` alongside each batch that flows through the system: 
```
WatermarkedAndCheckpointBatch {
  watermark: i64
  checkpoint: usize
  batch: Batch 
} 
```

I'm not sure I understand the idea for unordered batches in memory, but that likely needs a whole conversation around late data.